repos/system_upgrade/el7toel8/actors/ziplcheckbootentries/libraries/ziplcheckbootentries.py

```diff
@@ -0,0 +1,134 @@
+    """
+    Inhibits the upgrade if we are running on s390x and the bootloader configuration
+    contains multiple rescue boot entries.
+
+    A boot entry is recognized as a rescue entry when its title contains the `rescue` substring.
+
+    :param SourceBootloaderConfiguration bootloader_config: The configuration of the source boot loader.
+    """
+
+    # Keep the whole information about boot entries not just their count as
+    # we want to provide user with the details
+    rescue_entries = []
+    for boot_entry in bootloader_config.entries:
+        if is_rescue_entry(boot_entry):
+            rescue_entries.append(boot_entry)
+
+    if len(rescue_entries) > 1:
+        # Prepare the list of available rescue entries for user
+        rescue_entries_text = ''
+        for rescue_entry in rescue_entries:
+            rescue_entries_text += '{0}{1}'.format(FMT_LIST_SEPARATOR, rescue_entry.title)
+
+        summary = ('The Zipl configuration file {0} contains multiple rescue boot entries preventing migration '
+                   'to BLS. Problematic entries: {1}')
+
+        reporting.create_report([
+            reporting.Title('Multiple rescue boot entries present in the bootloader configuration.'),
+            reporting.Summary(summary.format(ZIPL_CONFIG_PATH, rescue_entries_text)),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.BOOT]),
+            reporting.Remediation(hint='Remove rescue boot entries from the configuration and leave just one.'),
+            reporting.Flags([reporting.Flags.INHIBITOR])
+        ])
+
+
+def extract_kernel_version(kernel_img_path):
+    """
+    Extracts the kernel version out of the given image path.
+
+    The extraction logic is designed to closely mimick the logic Zipl configuration to BLS
+    conversion script works, so that it is possible to identify the possible issues with kernel
+    images.
+
+    :param str kernel_img_path: The path to the kernel image.
+    :returns: Extracted kernel version from the given path
+    :rtype: str
+    """
+
+    # Mimick bash substitution used in the conversion script, see:
+    # https://github.com/ibm-s390-linux/s390-tools/blob/b5604850ab66f862850568a37404faa647b5c098/scripts/zipl-switch-to-blscfg#L168
+    if 'vmlinuz-' in kernel_img_path:
+        fragments = kernel_img_path.rsplit('/vmlinuz-', 1)
+        return fragments[1] if len(fragments) > 1 else fragments[0]
+
+    fragments = kernel_img_path.rsplit('/', 1)
+    return fragments[1] if len(fragments) > 1 else fragments[0]
+
+
+def inhibit_if_entries_share_kernel_version(bootloader_config):
+    """
+    Inhibits the upgrade if there are boot entries sharing the same kernel image version.
+
+    The logic of identification whether the images are the same mimics the zipl-switch-to-blscfg, as it fails
+    to perform the conversion if there are entries with the same kernel image.
```

Comment:
By mimicking the script aren't we opening doors to duplicating that script's validation logic in the leapp upgrade scripts?

Reply:
The script allows for being run into a temporary bls folder

```
--bls-directory=DIR     path to generate BLS files, defaults to /boot/loader/entries
--config-file=FILE      path to zipl configuration file, defaults to /etc/zipl.conf
```
Would it make sense to try this and inhibit on failure passing on the scripts output? This way the validation logic would remain in the zipl-switch-to-blscfg script and any new case wouldn't have to be added to leapp's logic IMO.

Reply:
The script is quite bad at explaining to the user why did it fail - it terminates with `File already exists.` and does not provide any further explanation about what caused the file to exist, which can be considered as a painfully bad user experience (no instructions on what exactly went wrong or how to fix it).